### DISCUSSION TITLE
Add multiple Authenticator apps for signon

### DISCUSSION THREAD

**Discussion by J-Cake** [community/community#60444] [Other Feature Feedback, Questions, & Ideas] [answered]
### Select Topic Area

Question

### Body

Hi folks, I'm a huge fan of 2FA apps and use then for everything. The issue is I'm lazy, and have a desktop program for this, so I can log into github from my PC without touching my phone. But it's not always accessible, and if I need to access GitHub when I'm away from my PC, I face difficulties. I'm wondering if there's a way to add multiple Authenticator apps to GitHub? I can add multiple 2FA methods, but not two of the same. I do understand the reasoning / technicalities behind it, having implemented something similar myself, but I would love the option to have multiple 2FA apps simultaneously. What might my options look like?

Best,
JCake

**Comment by jshtuazon2** [comment]
gmail.com

**Comment by CendioOssman** [comment]
Any response from GitHub on this? Now that 2FA is getting mandatory, flexibility and ease of use if even more important.

**Reply by Clemens85** [reply]
I am also quite disappointed... it is still not possible to add multilple authenticator apps (which is really useful if you have e.g. a second mobile phone).
You made MFA mandatory, then make it also more usable...

**Comment by joshuawalker** [comment]
When you're setting up (or reconfiguring) 2FA with an authenticator app, you can save the one-time password secret key to multiple authenticator apps (e.g. KeePassXC on your desktop and Microsoft Authenticator on your mobile device). Each app will generate the same TOTP code. You will need to save the secret key during setup: GitHub will not reveal your secret key again once you save your 2FA configuration.

**Reply by J-Cake** [reply]
This is a good suggestion, but to me this defeats the purpose of authenticator apps, in that the code is unique to the device. I mean true, this is a workaround, and happens to be exactly what I asked for, but true support for multiple Authenticator apps would be best.

**Comment by john-peterson** [comment]
So there is one pass key and one authenticator totp setup key

Should I save the setup key so I can enter it in more apps for safety?

I can't see the pass key so I can't manually save it anywhere

**Reply by john-peterson** [reply]
I have to save this topic because I am totally confused I have never in my life used this before

Like being forced to eat with chop sticks at forty five

**Reply by J-Cake** [reply]
> So there is one pass key and one authenticator totp setup key
> 
> Should I save the setup key so I can enter it in more apps for safety?
> 
> I can't see the pass key so I can't manually save it anywhere 

So I actually took some time to investigate this. I'm using three different apps; Microsoft Authenticator, Samsung Wallet and KDE Keysmith. Turns out this is exactly everything you need. GitHub will generate a QR code containing seed (not sure what you call it, but I call it a seed) which Samsung Wallet and Authenticator both understand to generate the next code. Crucially, if you scan the same code into the various apps, they will always generate the same OTP. For Keysmith, I had to decode the QR code manually because I'd already dismissed the prompt. But you can't retrieve the seed once it's been saved or scanned. So you kinda have to set them all up at the same time. What I did was take a screenshot of the QR code so I can reuse it. But be careful with that as that QR code allows you to access the GH account.

**Comment by jpanagos** [comment, marked as answer]
Just configured 2FA on two separate computers, both using the same type of method (browser authenticator app -- https://authenticator.cc/). The process is as follows:

- Login using your current method.
- Under https://github.com/settings/security, click on `Edit` for your preferred method.
- This will generate a secret key, treat it like a password, and save it _securely_. 
- Use this key to setup the authenticator on both devices.
- Click on the authenticator app icon, then the `pencil` icon, then the `+` sign, select `Manual Entry`, set `GitHub` as the issuer, then your secret key, and your username under the `Advanced` tab (I didn't change any other setting there).

You should be able to login without problems on both devices using the same type of 2FA method.

I cannot comment on other methods, but this worked for me and maybe it will help someone.

**Reply by cmette** [reply]
This is definitely not possible. I have tested it several times and tried it with various apps, FreeOTP, MicrosoftAuthenticstor etc. Only one end device can be registered. This is of course a big disadvantage!

**Reply by MarkoRamius** [reply]
_"And then, for the rest of your life, never run into that fucking Edit again."_ This.

**Reply by Adilius** [reply]
Just tried this and it worked. Successfully setup 2FA on two devices. Both of them generate the same code each time.

**Reply by jonaskoelker-jypo** [reply]
> saved into an independent file

Or just paste it into an xterm (notepad, etc.) while you input it in your MFA apps, then delete it and close whatever temporary storage space you kept it in, hoping that no program with friendly intentions stored the code on disk while you were doing your thing.

Apropos nothing, when switching from one device to another I recommend the following procedure:

1. Generate the shared secret key as described above
2. Input the key on the new device to create what most MFA apps call 'an account'
3. Type the verification code into GitHub.  This effectively switches to your new device.
4. Delete the GitHub account from your old MFA app
5. Create a new GitHub account in your old MFA app (using the key generated in step 1). This resurrects your old device as a clone of your new device, wrt. the GitHub MFA.
6. Visually confirm that the two MFA apps generate identical codes.

Don't do step 4 before step 3.  Maybe create a new GitHub account on the old device before deleting the old GitHub account, although I don't think this will have any effect: once you have verified the new secret key from at least one app, the old secret key (~= your old device) will no longer be accepted.

However you order the steps, **make sure you always have at least one working MFA code generator**, which in practice means "never delete whichever one you have verified most recently".

You can stop after step 3. Continuing to step 6 lets you use both the old and new devices interchangeably, e.g. during the period in which you switch everything else to your new device as well.

[a comment by an automated account is not shown]

**Comment by huan** [comment]
Yo GitHub crew, can we talk about the whole “one-authorizer-app-at-a-time” setup? Like, why can’t we just add a new one without having to boot the old one out?

Tech-wise, it’s not rocket science. Take Heroku, for example—they let you roll with two or more authorizer apps no sweat. I used to rock Authy there, and just recently, I added Apple Password without breaking a sweat.

This would be clutch for so many scenarios, so let’s rally and get this on the radar, yeah? Would love to see it happen!

**Comment by GlitterHorn** [comment]
Adding my voice to the rest here - I just had to do manual account recovery after forgetting to set up another device before wiping my previous one, and went to set up two OTP apps so I don't have to go through recovery again. I'd rather be able to set up multiple OTP apps, than require GitHub to use human labor to do manual account validations...

**Comment by J-Cake** [comment]
Hey, @J-Cake from two years later here: KeePass XC, and KeePass DX both support OTP in a KeePass 2 database.

 They can be synchronised to as many devices as you need. This is the setup I've been using recently and it works really well. 

Additionally, you can get the keys back out after adding it to the database

**Comment by evanblake1** [comment]
I get why you’d want multiple authenticator apps for flexibility, it feels similar to how a [Professional Event Management Company Oslo Norway ](https://pearllemonexperiences.com/event-management-company-oslo-norway/) coordinates backup plans so things keep running smoothly no matter the situation.
